targeting/geoip2/geoip2.go

```diff
@@ -123,6 +140,50 @@ func (g *GeoIP2) get(t geoType, db string) (*geoip2.Reader, error) {
 	return g.open(t, db)
 }
 
+func (g *GeoIP2) watchFiles() {
+	// Not worried about goroutines leaking because only one geoip2.New call is made in geodns.go (outside of testing)
+	ticker := time.NewTicker(1 * time.Minute)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				g.checkForUpdate()
+			default:
+				time.Sleep(12 * time.Second) // Sleep to avoid constant looping
+			}
+		}
+	}()
+}
+
+func (g *GeoIP2) checkForUpdate() {
+	// Iterate through each file, check modtime. If new, reload file
+	d := []*geodb{&g.country, &g.city, &g.asn} // Slice of pointers is kinda gross, but want to directly reference struct values (per const type)
+	for _, v := range d {
+		fi, err := os.Stat(v.fp)
+		if err != nil {
+			log.Printf("unable to stat DB file at %s :: %v", v.fp, err)
+			continue
+		}
+		if fi.ModTime().UTC().Unix() > v.lastModified {
+			g.mu.Lock()
+			e := v.db.Close()
+			if e != nil {
+				g.mu.Unlock()
```

Comment:
unlock with a defer (a couple lines up)?

Reply:
I'd much prefer it with defer.

The post you linked to could be summarized as "`defer` is function scoped" (duh). I've never seen those bugs in real code, but I've seen code with explicit unlock before returns grow or change and cause bugs from missing the unlock.

Using defer (and structuring the code so the lock is scoped to the function) avoids those bugs.

(I'm not suggesting it be done in this change, but at a glance if following my own advice the the get and open functions should be inverted so open() calls get and returns the cached value if one exists; that way open() could manage the lock instead of the awkward handling of it in get and open).

Reply:
Migrated to a new `reloadFile` function in 1942a4d330c1cd7fb3cdda6aa1b819f8c41c9c33